pkg/tso/keyspace_group_manager.go

```diff
@@ -495,39 +505,86 @@ func (kgm *KeyspaceGroupManager) deleteKeyspaceGroup(id uint32) {
+) (*AllocatorManager, error) {
+	if am := kgm.ams[keyspaceGroupID].Load(); am != nil {
+		ksg := kgm.ksgs[keyspaceGroupID].Load()
+		if ksg == nil {
+			return nil, kgm.genNotServedErr(errs.ErrGetAllocatorManager, keyspaceGroupID)
+		}
+		if _, ok := ksg.KeyspaceLookupTable[keyspaceID]; !ok {
+			return nil, kgm.genNotServedErr(errs.ErrGetAllocatorManager, keyspaceGroupID)
+		}
 		return am, nil
 	}
-	return nil, errs.ErrGetAllocatorManager.FastGenByArgs(
-		fmt.Sprintf("requested keyspace group with id %d %s by this host/pod", id, errs.NotServedErr))
+	return nil, kgm.genNotServedErr(errs.ErrGetAllocatorManager, keyspaceGroupID)
 }
 
 // GetElectionMember returns the election member of the given keyspace group
-func (kgm *KeyspaceGroupManager) GetElectionMember(id uint32) (ElectionMember, error) {
-	am, err := kgm.GetAllocatorManager(id)
+// TODO: support multiple keyspace groups for GetElectionMember
+func (kgm *KeyspaceGroupManager) GetElectionMember(
+	keyspaceID, keyspaceGroupID uint32,
+) (ElectionMember, error) {
+	if err := kgm.checkKeySpaceGroupID(keyspaceGroupID); err != nil {
+		return nil, err
+	}
+	am, err := kgm.GetAMWithMembershipCheck(keyspaceID, keyspaceGroupID)
 	if err != nil {
 		return nil, err
 	}
 	return am.getMember(), nil
 }
 
 // HandleTSORequest forwards TSO allocation requests to correct TSO Allocators of the given keyspace group.
-func (kgm *KeyspaceGroupManager) HandleTSORequest(id uint32, dcLocation string, count uint32) (pdpb.Timestamp, error) {
-	am, err := kgm.GetAllocatorManager(id)
+func (kgm *KeyspaceGroupManager) HandleTSORequest(
+	keyspaceID, keyspaceGroupID uint32,
+	dcLocation string, count uint32,
+) (ts pdpb.Timestamp, currentKeyspaceGroupID uint32, err error) {
+	if err := kgm.checkKeySpaceGroupID(keyspaceGroupID); err != nil {
+		return pdpb.Timestamp{}, keyspaceGroupID, err
+	}
+	am, err := kgm.GetAMWithMembershipCheck(keyspaceID, keyspaceGroupID)
 	if err != nil {
-		return pdpb.Timestamp{}, err
+		// The keyspace doesn't belong to this keyspace group, we should check if it belongs to any other
+		// keyspace groups, and return the correct keyspace group ID to the client.
+		for i := 0; i < int(mcsutils.MaxKeyspaceGroupCountInUse); i++ {
+			if ksg := kgm.ksgs[i].Load(); ksg == nil {
+				continue
+			} else if _, ok := ksg.KeyspaceLookupTable[keyspaceID]; ok {
+				return pdpb.Timestamp{}, ksg.ID, err
+			}
+		}
+		return pdpb.Timestamp{}, keyspaceGroupID, err
 	}
-	return am.HandleRequest(dcLocation, count)
+	ts, err = am.HandleRequest(dcLocation, count)
+	return ts, keyspaceGroupID, err
 }
 
 func (kgm *KeyspaceGroupManager) checkKeySpaceGroupID(id uint32) error {
 	if id < mcsutils.MaxKeyspaceGroupCountInUse {
 		return nil
 	}
 	return errs.ErrKeyspaceGroupIDInvalid.FastGenByArgs(
-		fmt.Sprintf("invalid keyspace group id %d which shouldn't >= %d", id, mcsutils.MaxKeyspaceGroupCountInUse))
+		fmt.Sprintf("invalid keyspace group id %d which shouldn't >= %d",
```

Comment:
I think the error message "invalid keyspace group id" has already been included in `ErrKeyspaceGroupIDInvalid`.

Reply:
Indeed. Changed to "fmt.Sprintf("%d which shouldn't >= %d", id, mcsutils.MaxKeyspaceGroupCountInUse))"